File: comparison/hof/hof.hs
#!/usr/bin/env runghc

-- Haskell functions can take functions as parameters
-- and return functions as return values.
-- A function that does either of those is called
-- a higher order function.

applyTwice :: (a -> a) -> a -> a
applyTwice f x = f (f x)

main :: IO ()
main = do
    print $ applyTwice (+3) 10

File: comparison/hof/hof.rs
#!/usr/bin/env rust-script

fn apply_twice<T>(
    f: T,
    x: i32,
) -> i32
where
    F: Fn(i32) -> i32,
{
    f(f(x))
}

fn main() -> () {
    let result = apply_twice(|x| x + 3, 10);
    println!("{}", result);
}
